### ISSUE TITLE
Is there a way to get the proof verifier to work for Ethereum MPT proofs?

### ISSUE THREAD

**Issue by kevincheng96** [paritytech/trie#141]
After playing around with this library, seems like the proof verifier may be incompatible with Ethereum merkle proofs generated thorough EIP-1186's eth_getProof method. Curious if I am missing something and if there is out-of-the-box support for Ethereum merkle proofs?

**Comment by cheme** [comment]
That is how it is use with ethereum https://github.com/openethereum/openethereum/blob/582bca385fedb1af682e989e5bcc6b3b2cf53028/crates/ethcore/src/state/account.rs#L652
Note that this is only for building the 'proof' field in the eip.

**Comment by lightyear15** [comment]
Hi @kevincheng96 ,
I am also interested in making trie-db proof verification function work with ethereum EIP-1186 proofs, did you manage to write a proper ``NodeCodec`` to make it work?

**Comment by lightyear15** [comment]
... Maybe @jimpo 
can give me an hint :smiling_face_with_tear:

**Comment by jimpo** [comment]
The `trie_db::proof` module was not designed to be compatible with EIP-1186, it was designed for space efficiency. From the module documentation: "The trie nodes are not included in their entirety as data which the verifier can compute for themself is omitted. In particular, the values of included keys and and hashes of other trie nodes in the proof are omitted."

It's somewhere in the Ethereum code but IIRC the way of verifying EIP-1186 proofs was to dump the nodes in the trie-db and do a lookup. Though EIP-1186 specifies an order and the verifier would have to check that separately.

API-wise, maybe it'd make sense to add a module for EIP-1186 proofs (full trie nodes), and rename or alias `trie_db::proof` to `trie_db::compact_proof` or something.

**Comment by lightyear15** [comment]
Thanks a lot @jimpo ,
I will explore your idea. 
I would rather create a module called ``extended_proof`` or a sub-module called ``eip1186`` though, so I won't break any existing code

**Comment by cheme** [comment]
> It's somewhere in the Ethereum code

see my previous comment :)

**Comment by seunlanlege** [comment]
So i found this crate: https://github.com/openethereum/openethereum/tree/582bca385fedb1af682e989e5bcc6b3b2cf53028/crates/db/patricia-trie-ethereum in open-ethereum but it looks like its no longer compatible with the latest version of this library.

What needs to be updated to make this compatible? Pseudo code is also fine 🙏🏿 

cc @cheme

**Comment by cheme** [comment]
Yes parity-ethereum looks a lot behind head, and project is not maintain anymore afaik.
The current master should work but it requires to upgrade all needed component:
- add a TrieLayout instance (should be pretty much this layout (this is what we test to try to remain eth compatibility) but replacing the codec with the rlp one.
- use builders to instanciate triedb and triedbmut.

Generally I would not hope that the compatibility with eth stays long term (mainly support the exstension nodes), since openeth stopped, I don't know of good reason in doing so. There is a point where this crate would be probably forked to a pure no-extension, limited rocksdb support one (first would require substrate to switch to rocksdb which could actually take some time:)

But with current state of code it should be supported.

**Comment by cheme** [comment]
forgot the link to the test layout : https://github.com/paritytech/trie/blob/48fcfa99c439949f55d29762e35f8793113edc91/test-support/reference-trie/src/lib.rs#L78

**Comment by seunlanlege** [comment]
> I did use it
https://github.com/ChorusOne/trie/blob/44a3c5e9c269111110537287aec8cc9050824446/trie-eip1186/test/src/eip1186.rs#L17

Hmm i guess my real question is: where's your `NodeCodec` implementation for Rlp?

**Comment by lightyear15** [comment]
Ah I see what you mean
no ``NodeCodec`` implementation.
The problem with trie-db is the way the proof is generated, not the way nodes are coded/decoded. 
So the ``trie-db::proof`` functions are of no use here

**Comment by seunlanlege** [comment]
Ended up implementing a `NodeCodec` for EIP-1186 here, if anyone ends up on this issue

https://github.com/polytope-labs/hyperbridge/blob/main/modules/trees/ethereum/README.md

**Comment by lightyear15** [comment]
@seunlanlege 

[This subcrate](https://github.com/paritytech/trie/tree/master/trie-eip1186)  has never been published, but it could've helped you
